octopus-server/src/test/java/octopus/teamcity/server/generic/RunbookRunStepTest.java

```diff
@@ -0,0 +1,164 @@
+package octopus.teamcity.server.generic;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import jetbrains.buildServer.serverSide.InvalidProperty;
+import octopus.teamcity.common.runbookrun.RunbookRunPropertyNames;
+import org.junit.jupiter.api.Test;
+
+class RunbookRunStepTest {
+
+  private final RunbookRunStep step = new RunbookRunStep();
+
+  @Test
+  public void describePropertiesProducesExpectedOutput() {
+    final String message = step.describeParameters(buildPropertiesMap());
+    assertThat("Runbook name: RunbookName\nProject name: ProjectName").isEqualTo(message);
+  }
+
+  @Test
+  public void validatePropertiesReturnsEmptyInvalidPropertiesListWithValidProperties() {
+    final List<InvalidProperty> invalidProperties = step.validateProperties(buildPropertiesMap());
+    assertThat(invalidProperties).isNotNull().hasSize(0);
+  }
+
+  @Test
+  public void validatePropertiesReturnsSingleInvalidPropertyNullRunbookName() {
+    final Map<String, String> properties = buildPropertiesMap();
+    properties.put(RunbookRunPropertyNames.RUNBOOK_NAME, null);
+
+    final List<InvalidProperty> invalidProperties = step.validateProperties(properties);
+    assertThat(invalidProperties)
+        .isNotNull()
+        .hasSize(1)
+        .flatExtracting(InvalidProperty::getPropertyName, InvalidProperty::getInvalidReason)
+        .containsExactly(
+            RunbookRunPropertyNames.RUNBOOK_NAME,
+            "Runbook name must be specified and cannot be whitespace.");
+  }
+
+  @Test
+  public void validatePropertiesReturnsSingleInvalidPropertyOnEmptyRunbookName() {
+    final Map<String, String> properties = buildPropertiesMap();
+    properties.put(RunbookRunPropertyNames.RUNBOOK_NAME, "");
+
+    final List<InvalidProperty> invalidProperties = step.validateProperties(properties);
+    assertThat(invalidProperties)
+        .isNotNull()
+        .hasSize(1)
+        .flatExtracting(InvalidProperty::getPropertyName, InvalidProperty::getInvalidReason)
+        .containsExactly(
+            RunbookRunPropertyNames.RUNBOOK_NAME,
+            "Runbook name must be specified and cannot be whitespace.");
+  }
+
+  @Test
```

Comment:
have you checkouted out @ParameterisedTest - you can probably get many of these tests "for free"

Reply:
vnext... now with parameterised testing